tower/src/util/future_service.rs

```diff
@@ -0,0 +1,163 @@
+use std::fmt;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_service::Service;
+
+/// Returns a new `FutureService` for the given future.
+///
+/// A `FutureService` allows you to treat a future that resolves to a service as a service. This
+/// can be useful for services that are created asynchronously.
+///
+/// # Example
+/// ```
+/// use tower::{service_fn, Service, ServiceExt};
+/// use tower::util::future_service;
+/// use std::convert::Infallible;
+///
+/// # fn main() {
+/// # async {
+/// let mut future_of_a_service = future_service(Box::pin(async {
+///     let svc = service_fn(|_req: ()| async { Ok::<_, Infallible>("ok") });
+///     Ok::<_, Infallible>(svc)
+/// }));
+///
+/// let svc = future_of_a_service.ready_and().await.unwrap();
```

Comment:
also, take it or leave it, but I might structure the example with creating the future and wrapping it with `future_service` on separate lines, so that it's clearer what's going on. something like this:
```suggestion
/// // A future which outputs a type implementing `Service`.
/// let future_of_a_service = async {
///     let svc = service_fn(|_req: ()| async { Ok::<_, Infallible>("ok") });
///     Ok::<_, Infallible>(svc)
/// };
///
/// // Wrap the future with a `FutureService`, allowing it to be used
/// // as a service without awaiting the future's completion:
/// let mut svc = future_service(Box::pin(future_of_a_service));
///
/// // Now, when we wait for the service to become ready, it will
/// // drive the future to completion internally.
/// let svc = svc.ready_and().await.unwrap();
```